src/ucs/datastruct/ptr_array.c

```diff
@@ -174,26 +195,40 @@ unsigned ucs_ptr_array_insert(ucs_ptr_array_t *ptr_array, void *value,
     /* Get the first item on the free list */
     index = ptr_array->freelist;
     ucs_assert(index != UCS_PTR_ARRAY_SENTINEL);
-    elem = &ptr_array->start[index];
 
-    /* Remove from free list */
-    ptr_array->freelist = ucs_ptr_array_freelist_get_next(*elem);
+    elem = &ptr_array->start[index];
 
-    /* Populate */
-    *placeholder_p = ucs_ptr_array_placeholder_get(*elem);
+    /* Remove from free list and Populate */
+    ptr_array->freelist = ucs_ptr_array_freelist_get_next_params(*elem,
+        &size_free_ahead);
     *elem = (uintptr_t)value;
+
     return index;
 }
 
 void ucs_ptr_array_remove(ucs_ptr_array_t *ptr_array, unsigned index,
                           uint32_t placeholder)
 {
     ucs_ptr_array_elem_t *elem = &ptr_array->start[index];
+    ucs_ptr_array_elem_t *next_elem;
+    uint32_t size_free_ahead;
 
     ucs_assert_always(!ucs_ptr_array_is_free(ptr_array, index));
-    *elem = UCS_PTR_ARRAY_FLAG_FREE;
-    ucs_ptr_array_placeholder_set(elem, placeholder);
-    ucs_ptr_array_freelist_set_next(elem, ptr_array->freelist);
+
+    if ( (index != UCS_PTR_ARRAY_SENTINEL) &&
+         (__ucs_ptr_array_is_free((ptr_array)->start[index+1])) ) {
+        next_elem = &ptr_array->start[index+1];
+        ucs_ptr_array_freelist_get_next_params(*next_elem,
+           &size_free_ahead);
+        size_free_ahead++;
+    }
+    else {
+        size_free_ahead = 1;
+    }
+
+    ucs_ptr_array_freelist_element_set_free(elem, size_free_ahead,
+       ptr_array->freelist);
+
     ptr_array->freelist = index;
```

Comment:
is this right? I think you mentioned the freelist only points to the beginning of a chunk of free elements, never to the middle. In this function if we remove an element surrounded by free ones - wouldn't it break the removal logic?